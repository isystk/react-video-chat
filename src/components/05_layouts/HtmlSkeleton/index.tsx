import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Header from '@/components/04_organisms/Header'
import SideMenu from '@/components/04_organisms/SideMenu'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { CssBaseline } from '@material-ui/core'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  main,
  title,
  description,
  isMenuOpen,
  setMenuOpen,
  children,
  ...props
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
    </Head>
    <Context.Provider value={main}>
      <CssBaseline />
      <Header isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
      <div>{children}</div>
      <SideMenu isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
    </Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()
  const [isMenuOpen, setMenuOpen] = useState(false)

  if (!main) return <></>

  const APP_NAME = process.env.APP_NAME

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : React.cloneElement(child)
  )
  if (!title) title = process.env.APP_NAME

  const description = `${APP_NAME} is an application that allows you to casually chat with your friends!`
  return presenter({
    main,
    title,
    description,
    isMenuOpen,
    setMenuOpen,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
