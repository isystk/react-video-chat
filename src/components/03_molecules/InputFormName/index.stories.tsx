import { Meta, Story } from '@storybook/react'
import React from 'react'
import InputFormName from './index'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

export default {
  title: '03_molecules/InputFormName',
  component: InputFormName,
} as Meta

const Template: Story = (props) => {
  const main = new MainService(() => ({}))

  return (
    <Context.Provider value={main}>
      <InputFormName />
    </Context.Provider>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
