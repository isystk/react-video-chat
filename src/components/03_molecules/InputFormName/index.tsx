import TextField from '@mui/material/TextField'
import React, { FC, useCallback, useContext, useEffect, useState } from 'react'
import Container from '@mui/material/Container'
import Button from '@mui/material/Button'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import Box from '@/components/01_atoms/Box'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import { FormControl, Typography } from '@mui/material'

/** InputFormNameProps Props */
export type InputFormNameProps = WithChildren
/** Presenter Props */
export type PresenterProps = InputFormNameProps & {
  main
  label
  name
  setName
  isComposed
  setIsComposed
  initializeLocalPeer
  disabled
}

/** Presenter Component */
const InputFormNamePresenter: FC<PresenterProps> = ({
  main,
  label,
  name,
  setName,
  isComposed,
  setIsComposed,
  initializeLocalPeer,
  disabled,
  ...props
}) => (
  <>
    <Container component="main">
      <Box title="ログイン">
        <FormControl className={styles.form}>
          <Typography component="h1" variant="h5">
            {label}を入力してください
          </Typography>
          <TextField
            autoFocus
            fullWidth
            label={label}
            margin="normal"
            name="name"
            onChange={(e) => setName(e.target.value)}
            onCompositionEnd={() => setIsComposed(false)}
            onCompositionStart={() => setIsComposed(true)}
            onKeyDown={async (e) => {
              if (isComposed) return
              if (e.target.value === '') return
              if (e.key === 'Enter') await initializeLocalPeer(e)
            }}
            required
            value={name}
            variant="outlined"
          />
          <Button
            className={styles.submit}
            color="primary"
            disabled={disabled}
            fullWidth
            onClick={async (e) => await initializeLocalPeer(e)}
            type="submit"
            variant="contained"
          >
            決定
          </Button>
        </FormControl>
      </Box>
    </Container>
  </>
)

/** Container Component */
const InputFormNameContainer: React.FC<
  ContainerProps<InputFormNameProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>

  const label = 'あなたの名前'
  const [disabled, setDisabled] = useState(true)
  const [name, setName] = useState('')
  const [isComposed, setIsComposed] = useState(false)

  useEffect(() => {
    const disabled = name === ''
    setDisabled(disabled)
  }, [name])

  const initializeLocalPeer = useCallback(
    async (e) => {
      e.persist()
      await main.setName(name)
      e.preventDefault()
    },
    [name, main]
  )

  if (main.self.name !== '') return <></>
  return presenter({
    children,
    main,
    label,
    name,
    setName,
    isComposed,
    setIsComposed,
    initializeLocalPeer,
    disabled,
    ...props,
  })
}

export default connect<InputFormNameProps, PresenterProps>(
  'InputFormName',
  InputFormNamePresenter,
  InputFormNameContainer
)
