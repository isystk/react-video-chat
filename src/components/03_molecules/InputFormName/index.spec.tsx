import React from 'react'
import renderer from 'react-test-renderer'
import InputFormName from './index'
import '@testing-library/jest-dom/extend-expect'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'

describe('InputFormName', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))

    const component = renderer.create(
      <Context.Provider value={main}>
        <InputFormName />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
