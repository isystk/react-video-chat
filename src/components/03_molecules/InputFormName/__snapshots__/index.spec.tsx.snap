// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFormName Match Snapshot 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiBox-root MuiBox-root-8 makeStyles-box-4"
  >
    <div
      className="MuiTypography-root makeStyles-title-5 MuiTypography-body1"
    >
      ログイン
    </div>
    <nav
      aria-label="breadcrumb"
      className="MuiTypography-root MuiBreadcrumbs-root makeStyles-breadcrumbs-6 MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      <ol
        className="MuiBreadcrumbs-ol"
      >
        <li
          className="MuiBreadcrumbs-li"
        >
          <a
            href="/"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            TOP
          </a>
        </li>
        <li
          aria-hidden={true}
          className="MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          className="MuiBreadcrumbs-li"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
          >
            ログイン
          </span>
        </li>
      </ol>
    </nav>
    <form
      className="makeStyles-form-7"
      noValidate={true}
    >
      <h1
        className="MuiTypography-root MuiTypography-h5"
      >
        あなたの名前
        を入力してください
      </h1>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onKeyDown={[Function]}
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
        >
          あなたの名前
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            name="name"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-11"
            >
              <span>
                あなたの名前
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-3 MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          決定
        </span>
      </button>
    </form>
  </div>
</main>
`;
